lgc/patch/PatchReadFirstLane.cpp

```diff
@@ -0,0 +1,477 @@
+  if (targetTransformInfoWrapperPass)
+    m_targetTransformInfo = &targetTransformInfoWrapperPass->getTTI(function);
+  assert(m_targetTransformInfo);
+
+  const bool changed = liftReadFirstLane(function);
+
+  return changed;
+}
+
+// =====================================================================================================================
+// Specify what analysis passes this pass depends on.
+//
+// @param [in,out] analysisUsage : The place to record our analysis pass usage requirements.
+void PatchReadFirstLane::getAnalysisUsage(AnalysisUsage &analysisUsage) const {
+  analysisUsage.addRequired<LegacyDivergenceAnalysis>();
+  analysisUsage.setPreservesCFG();
+}
+
+// =====================================================================================================================
+// Lift readfirstlanes in relevant basic blocks to transform VGPR instructions into SGPR instructions as many as
+// possible.
+//
+// @param [in,out] function : LLVM function to be run for readfirstlane optimization.
+bool PatchReadFirstLane::liftReadFirstLane(Function &function) {
+  // Collect the basic blocks with amdgcn_readfirstlane
+  // Build the map between initial readfirstlanes and their corrensponding blocks
+  DenseMap<BasicBlock *, SmallVector<Instruction *, 2>> blockInitialReadFirstLanesMap;
+  Module *module = function.getParent();
+  for (auto &func : *module) {
+    if (func.getIntrinsicID() == Intrinsic::amdgcn_readfirstlane) {
+      for (User *user : func.users()) {
+        Instruction *inst = cast<Instruction>(user);
+        if (inst->getFunction() != &function)
+          continue;
+        blockInitialReadFirstLanesMap[inst->getParent()].push_back(inst);
+      }
+      break;
+    }
+  }
+  bool changed = false;
+
+  // Lift readfirstlanes in each relevant basic block
+  for (auto blockInitialReadFirstLanes : blockInitialReadFirstLanesMap) {
+    BasicBlock *curBb = blockInitialReadFirstLanes.first;
+
+    // Step 1: Collect all instructions that "can be assumed uniform" with its divergent uses in a map
+    // (m_canAssumeUniformDivergentUsesMap)
+    collectAssumeUniforms(curBb, blockInitialReadFirstLanes.second);
+
+    // Step 2: Determine the best places to insert readfirstlane according to a heuristic
+    findBestInsertLocation(blockInitialReadFirstLanes.second);
+
+    // Step 3: Apply readFirstLane on all determined locations
+    assert(m_insertLocations.size() <= blockInitialReadFirstLanes.second.size());
+    BuilderBase builder(curBb->getContext());
+    for (auto inst : m_insertLocations) {
+      // Avoid to insert reduncant readfirstlane
+      if (auto intrinsic = dyn_cast<IntrinsicInst>(inst))
+        if (intrinsic->getIntrinsicID() == Intrinsic::amdgcn_readfirstlane)
+          continue;
+      if (areAllUserReadFirstLane(inst))
+        continue;
+
+      applyReadFirstLane(inst, builder);
+      changed = true;
+    }
+  }
```

Comment:
You can `m_canAssumeUniformDivergentUsesMap.clear()` at the end of each iteration, since each basic block is treated separately. It keeps the size of the structure smaller, and actually we **have** to clear it at the end anyway to avoid keeping dangling pointers.